core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageTest.java

```diff
@@ -306,6 +307,49 @@ void writeToFile(@TempDir Path tempDir) throws IOException {
         }
     }
 
+    @Test
+    void noopSubscribe() {
+        final StreamMessage<Integer> source = StreamMessage.of(1, 2, 3);
+        final List<Integer> collected = new ArrayList<>();
+        final StreamMessage<Integer> peeked = source.peek(collected::add);
+        final CompletableFuture<Void> cf = peeked.subscribe();
+        await().untilAsserted(() -> assertThat(cf.isDone()).isTrue());
+        assertThat(collected).isEqualTo(ImmutableList.of(1, 2, 3));
+    }
+
+    @Test
+    void noopSubscribe_abort() {
+        final StreamMessage<Integer> source = StreamMessage.of(1, 2, 3, 4, 5);
+        final List<Integer> collected = new ArrayList<>();
+        final DefaultStreamMessage<Integer> stream = new DefaultStreamMessage<>();
+        final StreamMessage<Integer> aborted = source
+                .peek(x -> {
+                    if (x == 3) {
+                        source.abort();
+                    } else {
+                        collected.add(x);
+                    }
+                });
+        await().untilAsserted(() -> assertThat(aborted.subscribe().isDone()).isTrue());
+        assertThat(collected).isEqualTo(ImmutableList.of(1, 2));
+    }
+
+    @Test
+    void noopSubscribe_error_thrown() throws Exception {
+        final StreamMessage<Integer> source = StreamMessage.of(1, 2, 3, 4, 5);
+        final List<Integer> collected = new ArrayList<>();
+        final StreamMessage<Integer> aborted = source
+                .peek(x -> {
+                    if (x == 3) {
+                        throw new RuntimeException();
+                    } else {
+                        collected.add(x);
+                    }
+                });
+        await().untilAsserted(() -> assertThat(aborted.subscribe().isDone()).isTrue());
+        assertThat(collected).isEqualTo(ImmutableList.of(1, 2));
+    }
+
```

Comment:
Could you also add a test case that makes sure the element is closed/released properly? You could:

1. Create a stream of `HttpData.wrap(ByteBuf)` with a pooled `ByteBuf`:
   ```
   ByteBuf buf1 = Unpooled.directBuffer(1);
   ByteBuf buf2 = Unpooled.directBuffer(1);
   HttpData data1 = HttpData.wrap(buf1);
   HttpData data2 = HttpData.wrap(buf2);
   StreamMessage<HttpData> source = StreamMessage.of(data1, data2);
   ```
3. Subscribe to the stream and wait until it's fully consumed by `NoopSubscriber`; and
4. Check if both `ByteBuf`s were released (i.e. `ByteBuf.refCnt() == 0`)

Reply:
Thank you for your kind feedback 🙇❤️  I'll also refer to netty's documentation!

Reply:
This has been resolved by @ikhoon's follow-up commit. :-)